libcudacxx/include/cuda/std/__cmath/isnan.h

```diff
@@ -158,11 +158,12 @@ template <class _Tp>
 #if _CCCL_HAS_FLOAT128()
 [[nodiscard]] _CCCL_API constexpr bool isnan(__float128 __x) noexcept
 {
-#  if defined(_CCCL_BUILTIN_ISNAN)
-  return _CCCL_BUILTIN_ISNAN(__x);
-#  else // ^^^ _CCCL_BUILTIN_ISNAN ^^^ / vvv !_CCCL_BUILTIN_ISNAN vvv
-  return ::cuda::std::__isnan_impl(__x);
-#  endif // ^^^ !_CCCL_BUILTIN_ISNAN ^^^
+  // __builtin_isnan is not efficient
+  if (!::cuda::std::__cccl_default_is_constant_evaluated())
+  {
+    NV_IF_TARGET(NV_IS_DEVICE, (return ::__nv_fp128_isnan(__x);)) // preserve NaN behavior even with optimization flags
+  }
```

Comment:
I am not sure thisis cromulent, we can add the device optimization but I want the `__is_nan_impl` remaining